framework-plugins/lisk-framework-chain-connector-plugin/src/active_validators_update.ts

```diff
@@ -83,15 +50,86 @@ export const calculateActiveValidatorsUpdate = (
 		certificateThreshold = validatorDataAtCertificate.certificateThreshold;
 	}
 
-	const activeBFTValidatorsUpdate = getActiveValidatorsDiff(
+	const activeValidatorsUpdate = getActiveValidatorsUpdate(
 		validatorDataAtLastCertificate.validators,
 		validatorDataAtCertificate.validators,
 	);
 
-	const activeValidatorsUpdate = activeBFTValidatorsUpdate.map(validator => ({
-		blsKey: validator.blsKey,
-		bftWeight: validator.bftWeight,
-	}));
-
 	return { activeValidatorsUpdate, certificateThreshold };
 };
+
+/**
+ * @see https://github.com/LiskHQ/lips/blob/main/proposals/lip-0053.md#computing-the-validators-update
+ */
+export const getActiveValidatorsUpdate = (
+	currentValidators: ActiveValidator[],
+	newValidators: ActiveValidator[],
+): ActiveValidatorsUpdate => {
+	const currentValidatorsMap = new utils.dataStructures.BufferMap<ActiveValidator>();
+	const allBLSKeysMap = new utils.dataStructures.BufferMap<Buffer>();
+	for (const v of currentValidators) {
+		currentValidatorsMap.set(v.blsKey, v);
+		allBLSKeysMap.set(v.blsKey, v.blsKey);
+	}
+
+	const validatorsUpdate: ActiveValidator[] = [];
+	const blsKeysUpdate: Buffer[] = [];
+	const newBLSKeys = new utils.dataStructures.BufferSet();
+	for (const v of newValidators) {
+		newBLSKeys.add(v.blsKey);
+		allBLSKeysMap.set(v.blsKey, v.blsKey);
+		const currentValidator = currentValidatorsMap.get(v.blsKey);
+		// if new validator does not exist, add to the blsKeysUpdate and validatorsUpdate
+		if (!currentValidator) {
+			blsKeysUpdate.push(v.blsKey);
+			validatorsUpdate.push(v);
+			continue;
+		}
+		// if currentValidator exist, but bft weight is different, then add to the validatorsUpdate
+		if (currentValidator.bftWeight !== v.bftWeight) {
+			validatorsUpdate.push(v);
+		}
+	}
+
+	for (const v of currentValidators) {
+		// We add the validator to the validatorsUpdate (with 0 BFT weight) if the BLS key is not in the newValidators.
+		if (!newBLSKeys.has(v.blsKey)) {
+			validatorsUpdate.push({
+				blsKey: v.blsKey,
+				bftWeight: BigInt(0),
+			});
+		}
+	}
+	blsKeysUpdate.sort((a, b) => a.compare(b));
+	validatorsUpdate.sort((a, b) => a.blsKey.compare(b.blsKey));
+
+	const bftWeightsUpdate = validatorsUpdate.map(v => v.bftWeight);
+	const allBLSKeys = allBLSKeysMap.values();
+	allBLSKeys.sort((a, b) => a.compare(b));
+
+	let bitmap = BigInt(0);
+	const bitmapSize = Number(BigInt(allBLSKeys.length + 7) / BigInt(8));
+	for (const validator of validatorsUpdate) {
+		const index = allBLSKeys.findIndex(k => k.equals(validator.blsKey));
+		bitmap |= BigInt(1) << BigInt(index);
+	}
+	const hexBitmap = bitmap.toString(16);
```

Comment:
I'm little confused with this part. So if we have a value of `bitmap` as `14n` after `bitmap |= BigInt(1) << BigInt(1)`, `bitmap |= BigInt(1) << BigInt(2)` and `bitmap |= BigInt(1) << BigInt(3)`, if we do `bitmap.toString(16) === 'e'` but if we do `bitmap.toString(2) === '1110'` , I suppose we wanted to use the later one with 2 base. But somehow the tests are passing so I wonder how it is working

Reply:
so to get 14n
we do `bitmap |= BigInt(1) << BigInt(1), bitmap |= BigInt(1) << BigInt(2) and bitmap |= BigInt(1) << BigInt(3)`
and hex `0e` and `1110` both are 14?

im not sure where you are confused

Reply:
I am confused in the code after it,
```
// prepend zero if length is odd number
	let bftWeightsUpdateBitmap = Buffer.from(
		hexBitmap.length % 2 === 0 ? hexBitmap : `0${hexBitmap}`,
		'hex',
	); // ===> based on hexBitmap='0e the bftWeightsUpdateBitmap = Buffer.from('0e', 'hex') , here I was thinking it would be `1110`?
	// if the result does not have enough length, pad zero to the left
	if (bftWeightsUpdateBitmap.length < bitmapSize) {
		bftWeightsUpdateBitmap = Buffer.concat([
			Buffer.alloc(bitmapSize - bftWeightsUpdateBitmap.length),
			bftWeightsUpdateBitmap,
		]);
	}

```

In the end if I understood it correctly it will give `bftWeightsUpdateBitmap = Buffer.from('0e', 'hex')`

Reply:
I checked the implementation once more and it seems correct